wallet-api/src/Wallet/API.hs

```diff
@@ -100,76 +104,97 @@ $(deriveEq1 ''EventTriggerF)
-        BlockHeightRange r    -> BlockHeightRange r
-        FundsAtAddress addr r -> FundsAtAddress addr r
+$(deriveEq1 ''AnnTriggerF)
+$(deriveOrd1 ''AnnTriggerF)
+$(deriveShow1 ''AnnTriggerF)
 
-instance Corecursive EventTrigger where
-    embed t = let f = EventTrigger . Fix in
-        case t of
-            And (EventTrigger l) (EventTrigger r) -> f $ And l r
-            Or (EventTrigger l) (EventTrigger r)  -> f $ Or l r
-            Not (EventTrigger n)                  -> f $ Not n
-            PAlways                               -> f PAlways
-            PNever                                -> f PNever
-            BlockHeightRange r                    -> f $ BlockHeightRange r
-            FundsAtAddress addr r                 -> f $ FundsAtAddress addr r
+getA :: AnnTriggerF a b -> a
+getA (AnnTriggerF (Compose (a, _))) = a
+
+newtype EventTrigger a = EventTrigger { getEventTrigger :: Fix (AnnTriggerF a) }
+    deriving (Eq, Ord, Show, Generic)
+
+getAnnot :: EventTrigger a -> a
+getAnnot (EventTrigger (Fix at)) = getA at
 
+type instance Base (EventTrigger a) = Compose ((,) a) EventTriggerF
+
+instance Recursive (EventTrigger a) where
+    project (EventTrigger (Fix (AnnTriggerF (Compose (v, k))))) = case k of
+        And l r               -> Compose (v, And (EventTrigger l) (EventTrigger r))
+        Or l r                -> Compose (v, Or (EventTrigger l) (EventTrigger r))
+        Not n                 -> Compose (v, Not (EventTrigger n))
+        PAlways               -> Compose (v, PAlways)
+        PNever                -> Compose (v, PNever)
+        BlockHeightRange r    -> Compose (v, BlockHeightRange r)
+        FundsAtAddress addr r -> Compose (v, FundsAtAddress addr r)
+
+instance Corecursive (EventTrigger a) where
+    embed (Compose (v, t)) = let f (a, b) = EventTrigger . Fix . AnnTriggerF . Compose $ (a, b) in
+        case t of
+            And (EventTrigger l) (EventTrigger r) -> f (v, And l r)
+            Or (EventTrigger l) (EventTrigger r)  -> f (v, Or l r)
+            Not (EventTrigger b)                  -> f (v, Not b)
+            PAlways                               -> f (v, PAlways)
+            PNever                                -> f (v, PNever)
+            BlockHeightRange r                    -> f (v, BlockHeightRange r)
+            FundsAtAddress addr r                 -> f (v, FundsAtAddress addr r)
 
 -- | `andT l r` is true when `l` and `r` are true.
-andT :: EventTrigger -> EventTrigger -> EventTrigger
-andT l = embed . And l
+andT :: EventTrigger () -> EventTrigger () -> EventTrigger ()
+andT l = embed . Compose . pure . And l
 
 -- | `orT l r` is true when `l` or `r` are true.
-orT :: EventTrigger -> EventTrigger -> EventTrigger
-orT l = embed . Or l
+orT :: EventTrigger () -> EventTrigger () -> EventTrigger ()
+orT l = embed . Compose . pure . Or l
 
 -- | `alwaysT` is always true
-alwaysT :: EventTrigger
-alwaysT = embed PAlways
+alwaysT :: EventTrigger ()
```

Comment:
It feels like it would be cleaner if this created a `Fix EventTriggerF`, rather than having the annotation layer in there unused.